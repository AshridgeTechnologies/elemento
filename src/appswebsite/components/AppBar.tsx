import * as React from 'react';
import Mui_AppBar from '@mui/material/AppBar';
import Box from '@mui/material/Box';
import Toolbar from '@mui/material/Toolbar';
import Typography from '@mui/material/Typography';
import IconButton from '@mui/material/IconButton';
import MenuIcon from '@mui/icons-material/Menu';
import UserMenu from '../../shared/UserMenu'

export default function AppBar() {
    return (
        <Box sx={{ flexGrow: 1 }}>
            <Mui_AppBar position="sticky">
                <Toolbar variant="dense">
                    <IconButton
                        size="large"
                        edge="start"
                        color="inherit"
                        aria-label="menu"
                        sx={{ mr: 2 }}
                    >
                        <MenuIcon />
                    </IconButton>
                    <Typography variant="h6" component="div" sx={{ fontSize: 16, flexGrow: 1 }}>
                        Elemento Apps
                    </Typography>
                    <UserMenu/>
                </Toolbar>
            </Mui_AppBar>
        </Box>
    );
}