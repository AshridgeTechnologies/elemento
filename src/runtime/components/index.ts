export {default as App} from './App'
export {default as Button} from './Button'
export {default as Icon} from './Icon'
export {default as Image} from './Image'
export {default as UserLogon} from './UserLogon'
export {default as Menu} from './Menu'
export {default as MenuItem} from './MenuItem'
export {default as NumberInput} from './NumberInput'
export {default as DateInput} from './DateInput'
export {default as TrueFalseInput} from './TrueFalseInput'
export {default as SpeechInput} from './SpeechInput'
export {default as Data} from './Data'
export {default as Calculation} from './Calculation'
export {default as Timer} from './Timer'
export {default as Collection} from './Collection'
export {default as Page} from './Page'
export {default as ComponentElement} from './ComponentElement'
export {default as Form} from './Form'
export {default as BaseFormState} from './FormState'
export {default as SelectInput} from './SelectInput'
export {default as TextElement} from './TextElement'
export {default as Layout} from './Layout'
export {default as Block} from './Block'
export {default as AppBar} from './AppBar'
export {default as ListElement} from './ListElement'
export {default as ListItem} from './ListItem'
export {default as ItemSet} from './ItemSet'
export {default as ItemSetItem} from './ItemSetItem'
export {default as TextInput} from './TextInput'
export {default as MemoryDataStore} from './MemoryDataStore'
export {default as FileDataStore} from './FileDataStore'
export {default as FileDataStoreImpl} from './FileDataStoreImpl'
export {default as IdbDataStoreImpl} from './IdbDataStoreImpl'
export {default as BrowserDataStore} from './BrowserDataStore'
export {default as FirestoreDataStore} from './FirestoreDataStore'
export {default as FirestoreDataStoreImpl} from './FirestoreDataStoreImpl'
export {default as ServerAppConnector} from './ServerAppConnector'
export {default as Adapter} from './Adapter'
