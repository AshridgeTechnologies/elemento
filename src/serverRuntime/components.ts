export {default as FirestoreDataStore} from './FirestoreDataStore'
export {default as Collection} from './Collection'

// NOTE: Must add the name of each component to ./names.ts