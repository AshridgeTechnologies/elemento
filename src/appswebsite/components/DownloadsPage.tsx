import React from 'react'
import {Box, Container, Link, Stack, Typography} from '@mui/material'
import AppBar from '../../shared/AppBar'
import {theme} from '../../shared/styling'
import {ThemeProvider} from '@mui/material/styles'

export default function LearnPage() {

    return <ThemeProvider theme={theme}>
        <Box display='flex' flexDirection='column' height='100%' width='100%'>
            <Box flex='0'>
                <AppBar title='Elemento Online - Downloads' userMenu={false}/>
            </Box>
            <Box flex='1' minHeight={0}>
                <Container>
                    <Stack maxWidth={600} spacing={3} mt={3}>
                        <Typography variant='h4'>Elemento Development Server</Typography>
                        <Typography>Please download the correct version for your operating system.</Typography>
                        <Typography>Your computer will complain that the app is not known, approved, signed etc.<br/>
                            We do not currently have the resources to jump through all the hoops required to get the app
                            approved on each operating system,
                            so you will just have to trust us here, and continue past all the warnings.
                        </Typography>
                        <Link href='/devServer/win/ElementoDevServer.exe'><Typography variant='h5'>Windows</Typography></Link>
                        <Link href='/devServer/macos/ElementoDevServer.zip'><Typography variant='h5'>Mac</Typography></Link>
                        <Link href='/devServer/linux/ElementoDevServer'><Typography variant='h5'>Linux</Typography></Link>
                    </Stack>
                </Container>
            </Box>
        </Box>
    </ThemeProvider>
}
