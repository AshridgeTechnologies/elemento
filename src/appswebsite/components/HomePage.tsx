import React from 'react'
import {Box, Container, Link, Stack, Typography} from '@mui/material'
import AppBar from '../../shared/AppBar'
import {theme} from '../../shared/styling'
import {ThemeProvider} from '@mui/material/styles'
import HomeIcon from '@mui/icons-material/Home'
import LightbulbIcon from '@mui/icons-material/LightbulbTwoTone'
import CreateIcon from '@mui/icons-material/CreateTwoTone'
import AodIcon from '@mui/icons-material/AodTwoTone'
import ArrowCircleRightTwoToneIcon from '@mui/icons-material/ArrowCircleRightTwoTone';

const primary = theme.palette.primary.main
const secondary = theme.palette.secondary.main

function Logo() {
    return <Box color={secondary} border='8px solid' borderRadius={4} width={108} height={104} paddingTop='10px' paddingX='3px'>
        <Stack>
            <Typography fontSize='100px' fontFamily='"Helvetica"' align='center' lineHeight={0.8}>El</Typography>
            <Typography fontSize='16px' lineHeight={1}>42</Typography>
        </Stack>
    </Box>
}

function SectionLink(props: {title: string, href: string, icon: any, children: any}) {
    const Icon = props.icon
    return <Box border='4px solid' borderColor={primary} borderRadius={4} width='300px' paddingY='10px' paddingX='16px'>
        <Link href={props.href} underline='none'>
            <Stack spacing={2}>
                <Box textAlign='center'><Icon sx={{ fontSize: 180 }} /></Box>
                <Link component='div' underline='hover' href={props.href} variant='h5' textAlign='center'>{props.title}</Link>
                <Typography fontSize='20px' color='#555'>{props.children}</Typography>
                <Box textAlign='center'><ArrowCircleRightTwoToneIcon sx={{ fontSize: 32, color: secondary }} /></Box>
            </Stack>
        </Link>
    </Box>
}

export default function HomePage() {

    return <ThemeProvider theme={theme}>
        <Box display='flex' flexDirection='column' height='100%' width='100%'>
            <Box flex='0'>
                <AppBar title='Elemento Apps - Home'/>
            </Box>
            <Box flex='1' minHeight={0} mt={2}>
                <Container maxWidth="lg">
                    <Stack spacing={2}>
                        <Stack direction='row'>
                            <Logo/>
                            <Stack ml={4}>
                                <Typography fontWeight='bold' variant='h3'  color={secondary} align='center'>Elemento</Typography>
                                <Typography variant='h4' color={secondary} align='center' mt={2}>
                                    Everything you need to make and use your own software
                                </Typography>
                            </Stack>
                        </Stack>
                        <Stack direction='row' justifyContent='space-between' alignItems='stretch' marginTop='30px'>
                            <SectionLink title='Learn' href='/learn' icon={LightbulbIcon}>
                                A great range of resources to teach you how to get started and create your first app with Elemento.
                            </SectionLink>
                            <SectionLink title='Create' href='/studio' icon={CreateIcon}>
                                Put your app together and preview it live in Elemento Studio
                            </SectionLink>
                            <SectionLink title='Use' href='/run' icon={AodIcon}>
                                Run your app - or go back to the apps you have already made.
                            </SectionLink>
                        </Stack>
                    </Stack>
                </Container>
            </Box>
        </Box>
    </ThemeProvider>
}
