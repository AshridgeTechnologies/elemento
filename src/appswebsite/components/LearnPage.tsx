import React from 'react'
import Toolbar from '@mui/material/Toolbar'
import {Box, Link, Stack, Typography} from '@mui/material'
import AppBar from '../../shared/AppBar'
import {theme} from '../../shared/styling'
import {ThemeProvider} from '@mui/material/styles'
import {TutorialRunner} from '../../tutorials/TutorialRunner'

export default function LearnPage() {

    return <ThemeProvider theme={theme}>
        <Box display='flex' flexDirection='column' height='100%' width='100%'>
            <Box flex='0'>
                <AppBar title='Elemento Apps - Learn' userMenu={false}/>
            </Box>
            <Box flex='1' minHeight={0}>
                <TutorialRunner/>
            </Box>
        </Box>
    </ThemeProvider>
}
