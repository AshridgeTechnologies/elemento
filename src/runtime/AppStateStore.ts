import SubscribableStore, {Callback, Id, UnsubscribeFn} from './SubscribableStore'
import {ComponentState} from './components/ComponentState'
import {unique} from 'radash'

export type StoredState = ComponentState<any>

export default class AppStateStore {

    constructor(private store: SubscribableStore = new SubscribableStore()) {
        this.subscribeToAllChanges()
    }

    get(id: string): StoredState {
        return this.store.get(id)
    }

    set<T extends StoredState>(id: string, item: T) {
        item.init(this.appStoreInterface(id), id)
        this.store.set(id, item)
    }

    updateIfChanged<T extends StoredState>(id: string, item: T) {
        const existingItem = this.get(id)
        const latestItem = existingItem ? existingItem.updateFrom(item) : item
        if (latestItem !== existingItem) {
            this.setDeferNotifications(id, latestItem)
        }
    }

    subscribe(id: string, callback: Callback): UnsubscribeFn {
        return this.store.subscribe(id, callback)
    }

    private subscribeToAllChanges() {
        this.store.subscribeAll( (ids: string[]) => this.updateParentItems(ids))
    }

    private updateParentItems(ids: Id[]) {
        const extractParent = (id: Id) => id.replace(/\.\w+$/, '')
        const parentIds = unique(ids.map(extractParent))
        parentIds.forEach( id => this.get(id)?.onChildStateChange() )
    }

    private setDeferNotifications<T extends StoredState>(id: string, item: T) {
        if (!this.store.isNotifyDeferred) {
            this.store.deferNotifications()
            setTimeout(() => this.store.sendNotifications(), 0)
        }
        this.set(id, item)
    }

    private appStoreInterface(id: string) {
        const store = this
        return {
            latest() {
                return store.get(id)
            },
            updateVersion(changes: object) {
                const updatedItem = store.get(id).withMergedState(changes)
                store.set(id, updatedItem)
            },
            getChildState(subPath: string) {
                return store.get(id + '.' + subPath)
            },
            getOrCreateChildState(subPath: string, item: StoredState) {
                store.updateIfChanged(id + '.' + subPath, item)
                return this.getChildState(subPath)
            },
            getApp() {
                const appId = id.split('.')[0]
                return store.get(appId)
            }
        }
    }
}
