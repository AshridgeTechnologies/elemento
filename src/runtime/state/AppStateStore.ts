import SubscribableStore, {type AllChangesCallback, type Callback, type Id, type UnsubscribeFn} from './SubscribableStore'
import {AppStateForObject} from '../components/ComponentState2'

type Props = { [p: string]: any }

export type StoredState = any

const placeholder = function() {const placeholder = 0}
placeholder.valueOf = () => undefined

export type MaybeInitable = { init?: (asi: AppStateForObject) => void }
export default class AppStateStore {

    constructor(private store: SubscribableStore = new SubscribableStore()) {}

    getRaw(id: string): StoredState {
        return this.store.get(id)
    }

    get(id: string): StoredState {
        const state = this.getRaw(id)
        return state && this.itemProxy(id, state)
    }

    getOrCreate<T extends MaybeInitable>(id: string, stateClass: new(...args: any[]) => T, stateProps: any): T {
        const existingState = this.getRaw(id)
        let targetState = existingState

        const initIfNeeded = (initialState: T) => {
            if (typeof initialState.init === 'function') {
                const store = this
                const asi: AppStateForObject = {
                    path: id,
                    latest() {
                        return store.get(id)
                    },
                    updateVersion(changes: object) {
                        const latestState = store.getRaw(id)
                        const updatedState = (latestState as any).withMergedState(changes)
                        store.setDeferNotifications(id, updatedState)
                    },
                    getChildState(subPath: string) {
                        return store.get(id + '.' + subPath)
                    }
                }
                initialState.init(asi)
            }
        }

        if (existingState === null || existingState === undefined) {
            const initialState = new stateClass(stateProps)
            initIfNeeded(initialState)
            targetState = initialState
            this.setDeferNotifications(id, initialState)
        } else if (!existingState._matchesProps(stateProps)) {
            const updatedState = (existingState as any).withProps(stateProps)
            initIfNeeded(updatedState)
            targetState = updatedState
            this.setDeferNotifications(id, updatedState)
        }

        return this.itemProxy(id, targetState)
    }

    subscribe(id: string, callback: Callback): UnsubscribeFn {
        return this.store.subscribe(id, callback)
    }

    subscribeAll(callback: AllChangesCallback): UnsubscribeFn {
        return this.store.subscribeAll(callback)
    }

    setDeferNotifications<T extends StoredState>(id: string, item: T) {
        if (!this.store.isNotifyDeferred) {
            this.store.deferNotifications()
            setTimeout(() => this.store.sendNotifications(), 0)
        }
        this.store.set(id, item)
    }

    private itemProxy(path: string, targetState: any) {
        const store = this
        const handler = {
            get(target: Props, property: string | symbol) {
                if (property === Symbol.toPrimitive) {
                    return target.valueOf
                }

                if (typeof property === 'symbol') {
                    return undefined
                }

                if (property === 'id') {
                    return (subId: Id) => `${path}.${subId}`
                }

                if (property in target) {
                    return target[property]
                }

                const childState = store.get(path + '.' + property)
                return childState ?? placeholder
            }
        }

        return new Proxy(targetState, handler)
    }
}
