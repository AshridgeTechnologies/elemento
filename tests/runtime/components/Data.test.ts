/**
 * @vitest-environment jsdom
 */
import {expect, test} from "vitest"
import {Data} from '../../../src/runtime/components/index'
import {snapshot, testAppInterfaceNew as testAppInterface, wrappedTestElementNew} from '../../testutil/testHelpers'
import {render} from '@testing-library/react'
import AppStateStore from '../../../src/runtime/state/AppStateStore'

const [data, appStoreHook] = wrappedTestElementNew(Data)

const stateAt = (path: string) => appStoreHook.stateAt(path)

const theStore = new AppStateStore()
let idSeq = 1
const createState = (props: object) => theStore.getOrCreate((idSeq++).toString(), Data.State, props)

test('Data element produces output with simple value',
    snapshot(data('app.page1.data1', {initialValue: 'Hi there!', display: true}))
)

test('Data element produces output with record value',
    snapshot(data('app.page1.data2', {initialValue: {a: 10, b: 'Bee1', c: true}, display: true}))
)

test('Data element produces empty output with default value for display', () => {
    const {container} = render(data('app.page1.height', {initialValue: 'Hi!'}))
    expect(container.innerHTML).toBe('')
})

test('Set replaces value', () => {
    const state = createState({initialValue: {a: 10, b: 'Bee1', c: true}})
    expect(state.value).toStrictEqual({a: 10, b: 'Bee1', c: true})

    state.Set({a:20, b:'Cee'})
    expect(state.latest().value).toStrictEqual({a:20, b:'Cee'})
})

test('Reset returns to initial value', () => {
    const state = createState({initialValue: {a: 10, b: 'Bee1', c: true}})
    state.updateState({value: {a:5, c:'Cee'}})
    expect(state.latest().value).toStrictEqual({a:5, c:'Cee'})

    state.Set({a:20, b:'Cee'})
    expect(state.latest().value).toStrictEqual({a:20, b:'Cee'})

    state.Reset()
    expect(state.latest().value).toStrictEqual({a: 10, b: 'Bee1', c: true})
})

test('Set can set array', () => {
    const state = createState({initialValue: []})
    expect(state.value).toStrictEqual([])

    state.Set(['a', 20])
    expect(state.latest().value).toStrictEqual(['a', 20])
})

test('Update changes only props given', () => {
    const state = createState({initialValue: {a: 10, b: 'Bee1', c: true}})
    expect(state.value).toStrictEqual({a: 10, b: 'Bee1', c: true})

    state.Update({a:20, b:'Cee'})
    expect(state.latest().value).toStrictEqual({a:20, b:'Cee', c: true})
})

test('Update can update an array', () => {
    const state = createState({initialValue: [10, 20, 30]})
    expect(state.value).toStrictEqual([10, 20, 30])

    state.Update({0:99, '2':'Cee'})
    expect(state.latest().value).toStrictEqual([99, 20, 'Cee'])
})

test('valueOf returns the value', () => {
    const initialValue = {a: 10, b: 'Bee1', c: true}
    const state = new Data.State({initialValue})
    expect(state.valueOf()).toBe(initialValue)
})

test('string conversion uses the value toString', () => {
    const initialValue = [1, 2, 3]
    const state = new Data.State({initialValue})
    expect('x' + state).toBe('x1,2,3')
})

test('properties of the value are copied onto the Data instance', () => {
    const initialValue = {a: 10, b: 'Bee1', c: true}
    const state = new Data.State({initialValue}) as any
    expect(state.a).toBe(10)
    expect(state.b).toBe('Bee1')
    expect(state.c).toBe(true)
})

test('no properties are copied onto the Data instance for a null value', () => {
    const state = new Data.State({initialValue: null})
    expect(state.value).toBe(null)
})

test('no properties are copied onto the Data instance for a primitive value', () => {
    const propNamesWithNull = Object.getOwnPropertyNames(new Data.State({initialValue: null}))
    const state = new Data.State({initialValue: 42})
    expect(state.value).toBe(42)
    const propNamesWith42 = Object.getOwnPropertyNames(state)
    expect(propNamesWith42).toStrictEqual(propNamesWithNull)
})

test('properties of the value are copied onto the Data instance after an update', () => {
    const initialValue = {a: 10, b: 'Bee1', c: true}
    // @ts-ignore
    const state = new Data.State({initialValue}).withState({value: {b: 'Bee2', c: false, d: '99'}}) as any
    expect(state.a).toBe(undefined)
    expect(state.b).toBe('Bee2')
    expect(state.c).toBe(false)
    expect(state.d).toBe('99')
})
