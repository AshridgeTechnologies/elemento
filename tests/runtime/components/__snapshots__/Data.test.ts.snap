// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Data element produces output with record value 1`] = `
<div
  id="app.page1.data2"
>
  <div>
    app.page1.data2
  </div>
  <code>
    {a: 10, b: 'Bee1', c: true}
  </code>
</div>
`;

exports[`Data element produces output with simple value 1`] = `
<div
  id="app.page1.data1"
>
  <div>
    app.page1.data1
  </div>
  <code>
    'Hi there!'
  </code>
</div>
`;
